scheduler/test/cook/test/simulator.clj

```diff
@@ -0,0 +1,571 @@
+
+            (log/info "Submitting batch")
+            ;; Submit new jobs
+            (doseq [job submission-batch]
+              (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
+              (submit-job mesos-datomic-conn job))
+            ;; Ensure peer has acknowledged the new jobs
+            (when (seq submission-batch)
+              (poll-until #(d/q '[:find ?j .
+                                      :in $ ?t
+                                      :where
+                                      [?j :job/submit-time ?t]]
+                                    (d/db mesos-datomic-conn)
+                                    ;; This relies on
+                                    ;; 1. Time is controlled
+                                    ;; 2. We increment time for each job submitted
+                                    (tc/to-date (t/now)))
+                        50
+                        60000))
+            (log/info "Batch submission complete")
+
+            ;; Rebalance
+            (log/info "Starting rebalance")
+            (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
+            (async/>!! rebalancer-trigger-chan rebalancer-complete-chan)
+            (async/<!! rebalancer-complete-chan)
+            (log/info "Rebalance complete")
+
+            ;; Request for jobs that are complete to have cook be notified
+            (log/info "Send completion status to scheduler")
+            (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
+            (async/>!! complete-trigger-chan flush-complete-chan)
+            (async/<!! flush-complete-chan)
+            
+            ;; Ensure mesos and cook state of running jobs matches
+            (poll-until #(= (set (map (comp str :instance/task-id)
+                                     (util/get-running-task-ents (d/db mesos-datomic-conn))))
+                            (set (map str
+                                     (keys (:task-id->task @state-atom)))))
+                       50
+                       60000
+                       #(let [cook-tasks (set (map (comp str :instance/task-id)
+                                                   (util/get-running-task-ents (d/db mesos-datomic-conn))))
+                              mesos-tasks (set (map str
+                                                    (keys (:task-id->task @state-atom))))]
+                          {:running-tasks-ents (count cook-tasks)
+                           :tasks-in-mesos (count mesos-tasks)
+                           :cook-minus-mesos (count (clojure.set/difference cook-tasks mesos-tasks))
+                           :mesos-minus-cook (count (clojure.set/difference mesos-tasks cook-tasks))})
+                       )
+            (log/info "Completion statuses sent")
+
+            ;; Request rank occurs
+            (log/info "Starting rank")
```

Comment:
Why does every simulation cycle trigger ranking? Shouldn't it be triggered based on whether enough simulated time has elapsed since last ranking?